### PR TITLE
Add Designs Unit Test

### PR DESCRIPTION
**Pull request by Gamer025**
## About The Pull Request
Adds a unit test for the design datum.
This unit tests checks for:
- Designs that have a custom ID but no name
- Designs that require materials to craft but don't provide anything
- Designs that requiere NO materials but give output
- Checks for any nanite designs that provide have no program_type
- Surgery designs with null or default ID/name/desc/surgery

Output looks like this:
image
The baddesign was a test datum I created
The shotgun stuff will go away once https://github.com/tgstation/tgstation/pull/55739 gets merged
Im not sure about the surgery warning, all surgery designs expect that single one have the "surgery" var set

## Why It's Good For The Game
Detects broken design datums

### CODE REVIEW

File: code/modules/unit_tests/designs.dm [hunk 1]
```diff
@@ -0,0 +1,38 @@
+/datum/unit_test/designs
+
+/datum/unit_test/designs/Run()
+	var/datum/design/defaultDesign = new /datum/design()
+	var/datum/design/nanites/defaultDesignNanites = new /datum/design/nanites()
+	var/datum/design/surgery/defaultDesignSurgery = new /datum/design/surgery()
```
Comment (Jared-Fogle) [inline code/modules/unit_tests/designs.dm:6]: Use snake_case variable names and use `allocate`, it'll qdel the datum when the test finishes.

Comment (Jared-Fogle) [reply code/modules/unit_tests/designs.dm:6]: Use snake_case variables everywhere else BTW.

File: code/modules/unit_tests/designs.dm [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+		var/datum/design/currentDesign = new path //Create an instance of each design
+		if (currentDesign.id == DESIGN_ID_IGNORE) //Don't check designs with ignore ID
+			continue
+		if (isnull(currentDesign.name) || currentDesign.name == defaultDesign.name) //Designs with ID must have non default/null Name
+			Fail("Design [currentDesign.type] has default or null name var but has an ID")
+		if ((!isnull(currentDesign.materials) && currentDesign.materials.len != 0) || (!isnull(currentDesign.reagents_list) && currentDesign.reagents_list.len != 0)) //Design requires materials
+			if ((isnull(currentDesign.build_path) || currentDesign.build_path == defaultDesign.build_path) && (isnull(currentDesign.make_reagents) || currentDesign.make_reagents == defaultDesign.make_reagents)) //Check if design gives any output
+				Fail("Design [currentDesign.type] requires materials but does not have have any build_path or make_reagents set")
+		else //Design requires no materials
+			if (!isnull(currentDesign.build_path) || !isnull(currentDesign.build_path)) //But gives stuff
```
Comment (Jared-Fogle) [inline code/modules/unit_tests/designs.dm:20]: `else if`

File: code/modules/unit_tests/designs.dm [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+	for(var/path in subtypesof(/datum/design))
+		if (ispath(path, /datum/design/nanites) || ispath(path, /datum/design/surgery)) //We are checking nanites and surgery design separatly later since they work differently
+			continue
+		var/datum/design/currentDesign = new path //Create an instance of each design
+		if (currentDesign.id == DESIGN_ID_IGNORE) //Don't check designs with ignore ID
+			continue
+		if (isnull(currentDesign.name) || currentDesign.name == defaultDesign.name) //Designs with ID must have non default/null Name
+			Fail("Design [currentDesign.type] has default or null name var but has an ID")
+		if ((!isnull(currentDesign.materials) && currentDesign.materials.len != 0) || (!isnull(currentDesign.reagents_list) && currentDesign.reagents_list.len != 0)) //Design requires materials
```
Comment (Jared-Fogle) [inline code/modules/unit_tests/designs.dm:16]: Use `LAZYLEN`, it'll compress this down a lot.

File: code/modules/unit_tests/designs.dm [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+	for(var/path in subtypesof(/datum/design/nanites))
+		var/datum/design/nanites/currentDesign = new path //Create an instance of each design
+		if (isnull(currentDesign.program_type) || currentDesign.program_type == defaultDesignNanites.program_type) //Check if the Nanite design provides a program
+			Fail("Nanite Design [currentDesign.type] does not have have any program_type set")
+
+	for(var/path in subtypesof(/datum/design/surgery))
+		var/datum/design/surgery/currentDesign = new path //Create an instance of each design
+		if (isnull(currentDesign.id) || currentDesign.id == defaultDesignSurgery.id) //Check if ID was not set
+			Fail("Surgery Design [currentDesign.type] has no ID set")
+		if (isnull(currentDesign.id) || currentDesign.name == defaultDesignSurgery.name ) //Check if name was not set
```
Comment (Jared-Fogle) [inline code/modules/unit_tests/designs.dm:32]:
```suggestion
		if (isnull(currentDesign.id) || currentDesign.name == defaultDesignSurgery.name) //Check if name was not set
```

**Review by Mothblocks — CHANGES_REQUESTED** [review]
Logic of the test is alright, just some pedantic concerns.

I'll revisit this when #55739 is merged.

File: code/modules/unit_tests/designs.dm (hunk 1 above)
Comment (Gamer025) [reply code/modules/unit_tests/designs.dm]: Changed everything to camel_case and allocate doesn't work as discussed (added comment at the top of run()

File: code/modules/unit_tests/designs.dm (hunk 2 above)
Comment (Gamer025) [reply code/modules/unit_tests/designs.dm]: Changed

File: code/modules/unit_tests/designs.dm (hunk 3 above)
Comment (Gamer025) [reply code/modules/unit_tests/designs.dm]: Changed

File: code/modules/unit_tests/designs.dm (hunk 4 above)
Comment (Gamer025) [reply code/modules/unit_tests/designs.dm]: Removed the whitespace

File: code/modules/unit_tests/designs.dm [hunk 5]
```diff
@@ -1,6 +1,7 @@
 /datum/unit_test/designs
 
 /datum/unit_test/designs/Run()
+//Can't use allocate because of bug with certain datums
```
Comment (Jared-Fogle) [inline code/modules/unit_tests/designs.dm:4]: Indent.

**Comment by Jared-Fogle** [comment]
merge with latest master

also, you should still try and figure out what's going on with the healing surgery

**Comment by Gamer025** [comment]
> merge with latest master
> 
> also, you should still try and figure out what's going on with the healing surgery, or just exclude it--iirc /healing is tend wounds (basic) and everything else is from upgrades

I was thinking about adding a blocklist for designs not to check but I think it would make the test more messy and in my opinion all the designs should really follow a common standard.

Looking at the surgery design again I think its actually only used for inheritance. (desc and icon_state)
I can nowhere in the code find a reference to /datum/design/surgery/healing or its id surgery_healing_base, so I think its not even available in game and even if it was it would have no use since the surgery var is not set.
Im thinking about just adding `surgery = /datum/surgery/healing` to it, any objections against that?

The surgery for "Tend Wounds (Bruise, Basic)" seems to be /datum/surgery/healing/brute/basic, which isn't part of any design.
I guess its just hard coded to be available, should probably be linked to a design which is part of the "Basic Medical equipment" which is researched by default.

**Comment by Jared-Fogle** [comment]
>  Im thinking about just adding surgery = /datum/surgery/healing to it, any objections against that?

Should be fine, CC @ExcessiveUseOfCobblestone if there's any reason that shouldn't work.

**Pull request merged**